mm2src/coins/utxo/utxo_common.rs

```diff
@@ -3604,56 +3604,48 @@ pub fn validate_payment<T: UtxoCommonOps>(
-                    continue;
-                },
-            };
-            if serialize(&tx).take() != tx_from_rpc.0
-                && serialize_with_flags(&tx, SERIALIZE_TRANSACTION_WITNESS).take() != tx_from_rpc.0
-            {
-                return MmError::err(ValidatePaymentError::WrongPaymentTx(format!(
-                    "Provided payment tx {:?} doesn't match tx data from rpc {:?}",
-                    tx, tx_from_rpc
-                )));
-            }
+        let tx_hash = tx.tx_hash();
 
-            let expected_output = TransactionOutput {
-                value: amount,
-                script_pubkey: Builder::build_p2sh(&dhash160(&expected_redeem).into()).into(),
-            };
+        let tx_from_rpc = retry_on_err!(coin
+            .as_ref()
+            .rpc_client
+            .get_transaction_bytes(&tx.hash().reversed().into())
+            .compat())
+        .repeat_every_secs(10.)
+        .attempts(4)
+        .inspect_err(move |e| error!("Error getting tx {tx_hash:?} from rpc: {e:?}"))
+        .await
+        .map_err(|timeout| timeout.error.map(ValidatePaymentError::from))?;
 
-            let actual_output = tx.outputs.get(output_index);
-            if actual_output != Some(&expected_output) {
-                return MmError::err(ValidatePaymentError::WrongPaymentTx(format!(
-                    "Provided payment tx output doesn't match expected {:?} {:?}",
-                    actual_output, expected_output
-                )));
-            }
+        if serialize(&tx).take() != tx_from_rpc.0
+            && serialize_with_flags(&tx, SERIALIZE_TRANSACTION_WITNESS).take() != tx_from_rpc.0
+        {
+            return MmError::err(ValidatePaymentError::WrongPaymentTx(format!(
+                "Provided payment tx {:?} doesn't match tx data from rpc {:?}",
+                tx, tx_from_rpc
+            )));
+        }
 
-            if let UtxoRpcClientEnum::Electrum(client) = &coin.as_ref().rpc_client {
-                if coin.as_ref().conf.enable_spv_proof && confirmations != 0 {
-                    client.validate_spv_proof(&tx, try_spv_proof_until).await?;
-                }
-            }
+        let expected_output = TransactionOutput {
+            value: amount,
+            script_pubkey: Builder::build_p2sh(&dhash160(&expected_redeem).into()).into(),
+        };
 
-            return Ok(());
+        let actual_output = tx.outputs.get(output_index);
+        if actual_output != Some(&expected_output) {
+            return MmError::err(ValidatePaymentError::WrongPaymentTx(format!(
+                "Provided payment tx output doesn't match expected {:?} {:?}",
+                actual_output, expected_output
+            )));
         }
+
+        if let UtxoRpcClientEnum::Electrum(client) = &coin.as_ref().rpc_client {
+            if coin.as_ref().conf.enable_spv_proof && confirmations != 0 {
+                client.validate_spv_proof(&tx, try_spv_proof_until).await?;
```

Comment:
Remembered that `try_spv_proof_until` was added because of how electrum servers behave after revising the related closed issue here https://github.com/KomodoPlatform/atomicDEX-API/issues/1239
I think we should leave this as it is :)

Reply:
Do you mean to refactor `validate_spv_proof` to return an error if `get_merkle_and_validated_header` fails and handle this error here
```rust
if let UtxoRpcClientEnum::Electrum(client) = &coin.as_ref().rpc_client {
    if coin.as_ref().conf.enable_spv_proof && confirmations != 0 {
        retry_in_err!(async { client.validate_spv_proof(&tx) }).await?;
    }
}
```

Reply:
Yes, that's what I meant.
But since electrums don't get blocks at the same time as stated in this closed issue https://github.com/KomodoPlatform/atomicDEX-API/issues/1239, if someone wants to get a transaction from one electrum and validate it using blocks from another electrum, then the retrying should be part of the implementation of `validate_spv_proof` for electrums, enabling waiting for the block of the transaction to appear.